addons/sale/tests/test_sale_order.py

```diff
@@ -196,3 +217,160 @@ def test_cost_invoicing(self):
+                line.product_id = self.service_deliver
+            with order.order_line.new() as line:
+                line.product_id = self.service_order
+            with order.order_line.new() as line:
+                line.product_id = self.product_deliver
+        order = sale_order.save()
+
+        # Check pricelist of the SO apply or not on order lines where pricelist contains formula that add 15% on the cost price
+        for line in order.order_line:
+            if line.product_id.categ_id in order.pricelist_id.item_ids.mapped('categ_id'):
+                for item in order.pricelist_id.item_ids.filtered(lambda l: l.categ_id == line.product_id.categ_id):
+                    discount = item.price_discount
+                    self.assertEquals(line.discount, discount, 'Pricelist of the SO should be applied on an order line')
+                self.assertEquals(line.price_unit, line.product_id.standard_price, 'Pricelist formula should be applied on the cost price')
+            else:
+                self.assertEquals(line.discount, 0.0, 'Pricelist of SO should not be applied on an order line')
+                self.assertEquals(line.price_unit, line.product_id.list_price, 'Unit price of order line should be a sale price as the pricelist not applied on the other category\'s product')
+
+    def test_sale_wtih_taxes(self):
+        """ Test SO with taxes """
+        # Create a tax with price included
+        tax_include = self.env['account.tax'].create({
+            'name': 'Tax with price include',
+            'amount': 10,
+            'price_include': True
+        })
+        # Create a tax with price not included
+        tax_exclude = self.env['account.tax'].create({
+            'name': 'Tax with no price include',
+            'amount': 10,
+        })
+
+        # Apply tax with price included on two products and tax with price not included on other two products
+        with Form(self.product_order) as product:
+            product.taxes_id.add(tax_include)
+        with Form(self.service_deliver) as product:
+            product.taxes_id.add(tax_include)
+        with Form(self.service_order) as product:
+            product.taxes_id.add(tax_exclude)
+        with Form(self.product_deliver) as product:
+            product.taxes_id.add(tax_exclude)
+
+        # Apply taxes on the sale order lines
+        sale_order = Form(self.sale_order)
+        with sale_order as order:
+            with order.order_line.edit(0) as line:
+                line.tax_id.add(tax_include)
+            with order.order_line.edit(1) as line:
+                line.tax_id.add(tax_include)
+            with order.order_line.edit(2) as line:
+                line.tax_id.add(tax_exclude)
+            with order.order_line.edit(3) as line:
+                line.tax_id.add(tax_exclude)
+        order = sale_order.save()
+
+        for line in order.order_line:
+            if line.tax_id == tax_include:
```

Comment:
rather check if the tax is "tax included" or not, than check if equals to this particular one.